### DISCUSSION TITLE
Is there any way to filter by whether a relation exists or not?

### DISCUSSION THREAD

**Discussion by jwld** [prisma/prisma#2772] [Q&A] [answered]
**Update: this is now covered quite thoroughly in the [docs](https://www.prisma.io/docs/orm/prisma-client/queries/relation-queries?utm_source=CLI&utm_medium=generator-warning#relation-filters)**

---

For the following model I want to run a query to get all the bookings that don't have any visits:
```
model Booking {
  visits Visit[]
}
```

Previously in prisma 1 I did this:
```
where: { visits_every: { id: null } } 
```

But I can't get the equivalent behaviour with prisma 2. The above always felt a bit hacky anyway.

Is there a definitive way to do this with prisma 2, or is it not supported? I guess I can resort to raw SQL if it's not.

**Comment by jwld** [comment]
And similarly for singular relations:
```
model Booking {
  worker Worker
}
```

Is the following valid in prisma 2, to get all bookings without a worker?
```
where: { worker: null }
```

**Reply by ryands17** [reply]
Hey @jwld :wave: 
You could do the same with Prisma 2 in the following manner:

```ts
where: { visits: { every: { id: null } } } 
```

**Reply by jwld** [reply]
Unfortunately it doesn't work - throws an error telling me to use `undefined` instead of `null`. But even doing that it's not filtering correctly.

```
Argument id for select.bookings.where.visits.every.id must not be null. Please use undefined instead.
```

**Reply by ryands17** [reply]
I have created an example similar to yours, as follows:

```prisma
model Post {
  id        String   @default(cuid()) @id
  published Boolean
  title     String
  content   String?
  authorId  Int?
  author    User?    @relation(fields: [authorId], references: [id])
  createdAt DateTime @default(now())
  updatedAt DateTime @updatedAt
}

model User {
  id        Int      @default(autoincrement()) @id
  email     String   @unique
  password  String
  name      String?
  posts     Post[]
  createdAt DateTime @default(now())
  updatedAt DateTime @updatedAt
}
```

And here I am querying to check for users who have no posts:

```ts
  const res = await prisma.user.findMany({
    where: {
      posts: {
        every: {
          authorId: null,
        },
      },
    },
  })
```
The above query works and returns the correct set of users. I am on version 2.0.0 of `@prisma/cli` and `@prisma/client`. Let me know if this works.

**Reply by jwld** [reply]
Ah right so you're checking whether the child's relation to the parent is null, rather than the id - interesting! That works for me too.

So the rule is that if you want to check whether a parent has children, you need to check the parent relation on the child?

**Reply by ryands17** [reply]
This would be one way of doing it. I would have to check if there's any other way :)

**Reply by jwld** [reply]
Thanks for the solution for now. Yea I'd be interested to know what the official recommendation is when filtering by whether a relation exists or not.

**Reply by jwld** [reply]
@ryands17 How would you do the following - I'm trying to get all the bookings that have a worker:

With prisma 1: 
```
where: { worker: { id_not: null } }
```

Is this sort of thing officially recommended or not? Because I've never seen it in the docs anywhere, but it's something we use a lot.

**Reply by ryands17** [reply]
@jwld It would be

```ts
where: { worker: { id: { not: null } } }
```

You could also find all these operators in the Prisma Client type definitions for reference.

**Reply by jwld** [reply]
Thanks, yea I'm aware of all these operators. This doesn't work however.

What I'm really trying to understand is whether this is the correct way to be doing this - is it a supported pattern that I can rely on not changing, or should I use raw sql instead?

**Reply by ryands17** [reply]
For the opposite that you have suggested above, I would suggest going for a raw query if it isn't possible as the reverse works. Let me check in the API again and let you know.

**Comment by ghost** [comment, marked as answer]
Posting this for anyone else looking at this in the future (since this is ranking high in Google search).

Here is how you can **filter on the existence of a one-to-one relation in prisma**:
```
// to find records that have a relation
where: {
  relation: { isNot: null }
}
// to find records that don't have a relation
where: {
  relation: { is: null }
}
```

Here is how you can **filter on the existence of a one-to-many relation in prisma**:
```
// to find records that do have at least 1 relation
where: {
  relation: { some: {} }
}
// to find records that don't have any relations
where: {
  relation: { none: {} }
}
```

**Reply by DaviCecilio** [reply]
Thanks!

**Reply by conioX** [reply]
Hi @iamarkadyt 

I Think is not working with me, this my code, schema, and response

```
 return await fastify.prisma.products.findMany({
            select: {
                id_product: true,
                category: {
                    select: {
                        category_name: true
                    }
                }
            },
            where: {
                category: {
                    isNot: null
                }
            }
        });
```


```
model products {
  id_product          Int                      @id @default(autoincrement())
  product_name        String?                  @db.VarChar(100)
  category_id         Int?                     @default(3)
  category            category?                @relation(fields: [category_id], references: [id_category])
}
```

```
    {
        "id_product": 1,
        "category": {
            "category_name": "Kategori 1"
        }
    },
    {
        "id_product": 9,
        "category": null
    },
```

still return null

**Reply by conioX** [reply]
i found it, instead null we can use {} object as our validation

**Reply by iway1** [reply]
This wouldn't work in my case. 

I had to do:

```tsx
await prisma.product.findFirst({
  where: {
    NOT: {
      category: null,
    }
  }
})
```

`isNot` doesn't accept `null` as a value in my case. It gave me an error and told me to pass "undefined" instead, but passing undefined didn't filter rows with no relation.

**Reply by AVAVT** [reply]
You saved my day @iway1 !

```
await prisma.item.findMany({
  where: {
    metadata: { isNot: null }
  }
})
```

Throw error: `item.metadata.isNot cannot be null. Please use undefined instead`
(`undefined` obviously doesn't work here, this error message is very unhelpful)


```
await prisma.item.findMany({
  where: {
    NOT: {
      metadata: null
    }
  }
})
```
^ this works as we expect

**Comment by laxmikanta415** [comment]
@iamarkadyt thanks mate for the answer. 
For anyone who want to refer it from database design to querying. 

Self relation with hierarchical data. 

Database schema:

**self-one to many relation:**

```
model Category {
  id       Int        @id @default(autoincrement())
  name     String?
  parentId Int?
  parentCategory  Category?  @relation("CategorySubcategories", fields: [parentId], references: [id])
  subCategories Category[] @relation("CategorySubcategories")

  @@map("category")
}
```

In three tiers:
- Category (level 1)
- subcategory 1 (level 2)
- subcategory 2 (level 3)


- Getting category level 1

```
this.prisma.category.findMany({
      distinct: ['name'],
      where: {
        parentId: null,
      },
    });
```
 has no parent.
    
- Getting subcategory 1 (level 2) :

```
 this.prisma.category.findMany({
      include: {
        subCategories: {
          select: {
            name: true,
          },
        },
        parentCategory: {
          select: {
            name: true,
          },
        },
      },
      where: {
        parentCategory: {
          isNot: null,
        },
        subCategories: { some: {} },
      },
    });
```
has parent and child.


- subcategory 2 (level 3)

```
this.prisma.category.findMany({
      include: {
        subCategories: true,
      },
      where: {
        parentCategory: {
          isNot: null,
        },
        subCategories: {
          none: {},
        },
      },
    });

```
has parent but no child.

**Reply by MorenoMdz** [reply]
None of this work inside an OR clause right? For some reason using `is: null` at the root where object works, but if I use it inside an OR array it errors with
```ts
Object literal may only specify known properties, and 'is' does not exist in type 'ModelWhereInput'.ts(2353)
```

**Comment by andrewmclagan** [comment]
This works but does not change the return type, it is still `Model | null` where it should be `Model`...

**Reply by olalonde** [reply]
Is there a work around for this?

**Reply by Magnuti** [reply]
https://github.com/prisma/prisma/issues/20348

**Comment by ludralph** [comment]
Hi @jwld,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
